### PR TITLE
Network status timeline

### PR DESCRIPTION
**Pull request by dylanbronson**
Adding NetworkStatusTimeline class to keep the timeline of the NetworkStatus. This will be used when we check whether a DirectMethod TestOperationResult occurred while online or offline. 
Also added unit tests for NetworkStatusTimeline.

### CODE REVIEW

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    using Newtonsoft.Json;
+
+    class NetworkStatusTimeline
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger(nameof(NetworkStatusTimeline));
+
+        readonly List<NetworkControllerTestResult> networkControllerTestResults;
+        readonly double tolerancePeriodInMilliseconds;
+
+        public static async Task<NetworkStatusTimeline> CreateNetworkStatusTimeline(ITestResultCollection<TestOperationResult> networkControllerCollection, double tolerancePeriodInMilliseconds)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
        public static async Task<NetworkStatusTimeline> CreateNetworkStatusTimeline(ITestResultCollection<TestOperationResult> networkControllerStatuses or networkControllerResults, TimeSpan tolerancePeriodInMilliseconds)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            List<NetworkControllerTestResult> networkControllerTestResults = new List<NetworkControllerTestResult>();
+            while (await networkControllerCollection.MoveNextAsync())
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion

            while (await networkControllerCollection.MoveNextAsync())
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                    r =>
+                    {
+                        networkControllerTestResults.Add(r);
+                    });
+            }
+
+            return new NetworkStatusTimeline(networkControllerTestResults, tolerancePeriodInMilliseconds);
+        }
+
+        NetworkStatusTimeline(List<NetworkControllerTestResult> networkControllerTestResults, double tolerancePeriodInMilliseconds)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
        NetworkStatusTimeline(IReadonlyList<NetworkControllerTestResult> networkControllerTestResults, double tolerancePeriodInMilliseconds)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        public NetworkControllerStatus GetNetworkControllerStatusAt(DateTime dateTime)
+        {
+            // Return network controller status at given time
+            NetworkControllerStatus networkControllerStatus = NetworkControllerStatus.Enabled;
+            for (int i = 0;  i < this.networkControllerTestResults.Count; i++)
+            {
+                // If given DateTime is temporally after a DateTime on the list, set the networkStatus to that
+                if (dateTime >= this.networkControllerTestResults[i].CreatedAt)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: First you can't assume given networkControllerTestRsults is sorted.  Second, I think you should try to find the largest created at that is less than given dateTime; and you need to throw exception if given datetime is less than all network controller test result created at timestamp.

Comment (philipktlin) [reply test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Also, I think you should combine IsWithinTolerancePeriod with this method and return both NetworkControllerStatus and IsWithinTolerancePeriod boolean value.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            return networkControllerStatus;
+        }
+
+        public bool IsWithinTolerancePeriod(DateTime dateTime)
+        {
+            // Return true if given dateTime is within a tolerance period of one of the NetworkControllerTestResults
+            foreach (NetworkControllerTestResult networkControllerTestResult in this.networkControllerTestResults)
+            {
+                // If dateTime is after a TestResult's CreatedAt time, but before that TestResult's CreatedAt time + tolerance period, then dateTime is within the tolerance period
+                if (dateTime > networkControllerTestResult.CreatedAt && dateTime <= networkControllerTestResult.CreatedAt.Add(this.tolerancePeriod))
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Let's discuss tomorrow.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs (hunk 4 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Combined the two methods and now not assuming order

File: test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.ModuleUtil.NetworkController
+{
+    public enum NetworkControllerOperation
+    {
+        SettingRule,
```
Comment (philipktlin) [inline test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs]: Updating/Changing

File: test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.ModuleUtil.NetworkController
+{
+    public enum NetworkControllerOperation
```
Comment (philipktlin) [inline test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs]: should this be NetworkControllerOperationStatus? which will have "InProgress" and "Finished"?  Because it looks odd as we just do 1 operation and we report before and after, therefore the enum name should not be "Operation".

Comment (philipktlin) [reply test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs]: I know this is something existing; I raise this as you touched this file, it is your call if you want to change it.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                Option.None<TwinTestResult>();
+            }
+
+            Logger.LogDebug($"Deserializing for source {current.Source} result: {current.Result} {current.Type}");
+            NetworkControllerTestResult twinTestResult = JsonConvert.DeserializeObject<NetworkControllerTestResult>(current.Result);
+            return Option.Some(twinTestResult);
+        }
+    }
+
+    public class SortNetworkControllerTestResultHelper : IComparer<NetworkControllerTestResult>
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Start with noun.
```suggestion
    public class NetworkTimelineResultComparer : IComparer<NetworkControllerTestResult>
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                Option<NetworkControllerTestResult> networkControllerTestResult = GetNetworkControllerTestOperationResult(networkControllerTestOperationResults.Current);
+
+                networkControllerTestResult.ForEach(
+                    r =>
+                    {
+                        networkControllerTestResults.Add(r);
+                    });
+            }
+
+            networkControllerTestResults.Sort(new SortNetworkControllerTestResultHelper());
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Call Sort method inside constructor

File: test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs (hunk 7 above)
Comment (dylanbronson) [reply test/modules/ModuleLib/NetworkController/NetworkControllerOperation.cs]: I see your point. Let's change it in another PR, though

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        NetworkStatusTimeline(List<NetworkControllerTestResult> networkControllerTestResults, TimeSpan tolerancePeriod)
+        {
+            networkControllerTestResults.Sort(new NetworkTimelineResultComparer());
+            this.networkControllerTestResults = networkControllerTestResults;
+            this.tolerancePeriod = tolerancePeriod;
+        }
+
+        private void ValidateNetworkControllerTestResults()
+        {
+            for (int i = 0; i < this.networkControllerTestResults.Count; i++)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
            for (int i = 0; i < this.networkControllerTestResults.Count; i+=2)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                        throw new InvalidOperationException("Test result SettingRule and following RuleSet do not match NetwokControllerStatuses");
+                    }
+                    i++;
+                }
+                else
+                {
+                    throw new InvalidOperationException("Expected SettingRule");
+                }
+            }
+            this.testResultsValidated = true;
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion

            this.testResultsValidated = true;
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        private void ValidateNetworkControllerTestResults()
+        {
+            for (int i = 0; i < this.networkControllerTestResults.Count; i++)
+            {
+                NetworkControllerTestResult curr = this.networkControllerTestResults[i];
+                if (NetworkControllerOperation.SettingRule.Equals(curr.Operation))
+                {
+                    if (i + 1 >= this.networkControllerTestResults.Count || !NetworkControllerOperation.RuleSet.Equals(this.networkControllerTestResults[i + 1].Operation))
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: split this if statement into 2; and throw different exception messages.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+            this.networkControllerTestResults = networkControllerTestResults;
+            this.tolerancePeriod = tolerancePeriod;
+        }
+
+        private void ValidateNetworkControllerTestResults()
+        {
+            for (int i = 0; i < this.networkControllerTestResults.Count; i++)
+            {
+                NetworkControllerTestResult curr = this.networkControllerTestResults[i];
+                if (NetworkControllerOperation.SettingRule.Equals(curr.Operation))
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: invert this if statement and throw exception immediately.  I think this is more readable.
e.g.
if (condition1)
    throw ...
if (condition2)
    throw ...
...

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                }
+                else
+                {
+                    throw new InvalidOperationException("Expected SettingRule");
+                }
+            }
+            this.testResultsValidated = true;
+        }
+
+        public (NetworkControllerStatus, bool) GetNetworkControllerStatusAndWithinToleranceAt(DateTime dateTime)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: use named tuple
```suggestion
        public (NetworkControllerStatus networkStatus, bool withinTolerance) GetNetworkControllerStatusAt(DateTime dateTime)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            if (this.testResultsValidated == false)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
            if (!this.testResultsValidated)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+            {
+                this.ValidateNetworkControllerTestResults();
+            }
+            // Return network controller status at given time
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: line break
```suggestion

            // Return network controller status at given time
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs (hunk 15 above)
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Move this if statement inside ValidateNetworkControllerTestResults method.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+            NetworkControllerStatus networkControllerStatus = NetworkControllerStatus.Unknown;
+            bool isWithinTolerancePeriod = false;
+            for (int i = 0;  i < this.networkControllerTestResults.Count; i++)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
            for (int i = 0;  i < this.networkControllerTestResults.Count; i+=2)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+                NetworkControllerTestResult curr = this.networkControllerTestResults[i];
+                if (dateTime <= curr.CreatedAt)
+                {
+                    break;
+                }
+
+                networkControllerStatus = curr.NetworkControllerStatus;
+                NetworkControllerTestResult next = this.networkControllerTestResults[i + 1];
+                isWithinTolerancePeriod = dateTime >= curr.CreatedAt && dateTime <= next.CreatedAt.Add(this.tolerancePeriod);
+                i++;
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: remove it.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+            {
+                NetworkControllerTestResult curr = this.networkControllerTestResults[i];
+                if (dateTime <= curr.CreatedAt)
+                {
+                    break;
+                }
+
+                networkControllerStatus = curr.NetworkControllerStatus;
+                NetworkControllerTestResult next = this.networkControllerTestResults[i + 1];
+                isWithinTolerancePeriod = dateTime >= curr.CreatedAt && dateTime <= next.CreatedAt.Add(this.tolerancePeriod);
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
                isWithinTolerancePeriod = dateTime > curr.CreatedAt && dateTime <= next.CreatedAt.Add(this.tolerancePeriod);
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+                i++;
+            }
+
+            if (NetworkControllerStatus.Unknown.Equals(networkControllerStatus))
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Currently test should start to run with network enabled; after n mins, network controller start to change network status. Therefore it will throw this exception.   I think you can either hard-code initial network status (line 81) to online or have an argument in constructor to get initial network status.

Comment (philipktlin) [reply test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Please add test to cover this line.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+                    });
+            }
+
+            networkControllerTestResults.Sort(new SortNetworkControllerTestResultHelper());
+            return new NetworkStatusTimeline(networkControllerTestResults, tolerancePeriod);
+        }
+
+        NetworkStatusTimeline(IReadOnlyList<NetworkControllerTestResult> networkControllerTestResults, TimeSpan tolerancePeriod)
+        {
+            this.networkControllerTestResults = networkControllerTestResults;
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Validate 1. results not null using PreConditions.CheckNotNull, 2. results count cannot be 0, 3. results count cannot be odd.

File: test/modules/Modules.Test/TestResultCoordinator/NetworkStatusTimelineTest.cs [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+                            new DateTime(2020, 1, 1, 9, 10, 15, 13),
+                            new DateTime(2020, 1, 1, 9, 10, 20, 10),
+                            new DateTime(2020, 1, 1, 9, 10, 20, 13),
+                            new DateTime(2020, 1, 1, 9, 10, 25, 10),
+                            new DateTime(2020, 1, 1, 9, 10, 25, 13)
+                        },
+                        new[] { "SettingRule", "SettingRule", "SettingRule", "RuleSet", "SettingRule", "RuleSet", "SettingRule", "RuleSet" }
+                    }
+                };
+        [Theory]
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/NetworkStatusTimelineTest.cs]: line break
```suggestion

        [Theory]
```

File: test/modules/Modules.Test/TestResultCoordinator/NetworkStatusTimelineTest.cs [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+                            new DateTime(2020, 1, 1, 9, 10, 25, 10),
+                            new DateTime(2020, 1, 1, 9, 10, 25, 13)
+                        },
+                        new[] { "SettingRule", "RuleSet", "SettingRule", "RuleSet", "SettingRule", "RuleSet", "SettingRule", "RuleSet" }
+                    }
+                };
+        public static IEnumerable<object[]> GetInvalidCreateReportData =>
+            new List<object[]>
+                {
+                    new object[]
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/NetworkStatusTimelineTest.cs]: Add more tests; for example empty results and odd number of results.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+                        networkControllerTestResults.Add(r);
+                    });
+            }
+
+            return new NetworkStatusTimeline(networkControllerTestResults, tolerancePeriod, initialNetworkControllerStatus);
+        }
+
+        NetworkStatusTimeline(List<NetworkControllerTestResult> networkControllerTestResults, TimeSpan tolerancePeriod, NetworkControllerStatus initialNetworkControllerStatus)
+        {
+            if (networkControllerTestResults.Count == 0)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
            if (networkControllerTestResults?.Count == 0)
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                throw new InvalidOperationException("Network Controller Test Results is empty.");
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]:
```suggestion
                throw new ArgumentException("Network Controller Test Results is empty.");
```

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+            networkControllerTestResults.Sort(new NetworkTimelineResultComparer());
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Do sorting against class variable, but not argument.

File: test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs (hunk 24 above)
Comment (philipktlin) [reply test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs]: Please also check if networkControllerTestResults.Count is odd number
